### PR TITLE
tilt: implement synclet

### PR DESCRIPTION
**Pull request by landism**

### CODE REVIEW

File: internal/synclet/synclet.go [hunk 1]
```diff
@@ -19,8 +23,70 @@ func NewSynclet(dcli build.DockerClient) *Synclet {
 	return &Synclet{dcli: dcli}
 }
 
-func (s Synclet) GetContainerIdForPod(podId string) (string, error) {
-	return "", errors.New("GetContainerIdForPod not implemented")
+const pauseCmd = "/pause"
+
+func (s Synclet) GetContainerIdForPod(ctx context.Context, podId string) (string, error) {
```
Comment (maiamcc) [inline internal/synclet/synclet.go]: should be using k8s.PodID wherever possible -- it IS just a string, but (according to nick) a good way to make args transparent and enforce that we're passing not only a string but a VALID string...?

File: internal/synclet/synclet.go [hunk 2]
```diff
@@ -19,8 +23,70 @@ func NewSynclet(dcli build.DockerClient) *Synclet {
...
+	// What?? No actual matches??!
+	return "", fmt.Errorf("no real containers -- all were '/pause' containers")
+}
+
+func (s Synclet) writeFiles(ctx context.Context, containerId string, tarArchive []byte) error {
+	if tarArchive == nil {
+		return nil
+	}
+
+	err := s.dcli.CopyToContainerRoot(ctx, containerId, bytes.NewBuffer(tarArchive))
```
Comment (maiamcc) [inline internal/synclet/synclet.go]: why not just `return s.dcli.CopyTo...`?

File: internal/synclet/synclet.go [hunk 3]
```diff
@@ -19,8 +23,70 @@ func NewSynclet(dcli build.DockerClient) *Synclet {
...
+	return err
+}
+
+func (s Synclet) rmFiles(ctx context.Context, containerId string, filesToDelete []string) error {
```
Comment (maiamcc) [inline internal/synclet/synclet.go]: again, should take k8s.ContainerId probs?

File: internal/synclet/synclet.go [hunk 4]
```diff
@@ -19,8 +23,70 @@ func NewSynclet(dcli build.DockerClient) *Synclet {
...
+	if len(filesToDelete) == 0 {
+		return nil
+	}
+
+	id := k8s.ContainerID(containerId)
+	cmd := model.Cmd{Argv: append([]string{"rm"}, filesToDelete...)}
+
+	err := s.dcli.ExecInContainer(ctx, id, cmd)
```
Comment (maiamcc) [inline internal/synclet/synclet.go]: `return s.dcli.Exec...`

File: internal/synclet/synclet.go [hunk 5]
```diff
@@ -19,8 +23,70 @@ func NewSynclet(dcli build.DockerClient) *Synclet {
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s Synclet) restartContainer(ctx context.Context, containerId string) error {
+	err := s.dcli.ContainerRestartNoWait(ctx, containerId)
```
Comment (maiamcc) [inline internal/synclet/synclet.go]: `return s.dcli...`

File: internal/synclet/synclet.go [hunk 6]
```diff
@@ -30,10 +96,24 @@ func (s Synclet) UpdateContainer(
 	filesToDelete []string,
 	commands []model.Cmd) error {
 
-	fmt.Println("container:", containerId)
-	fmt.Println("bytes in tar:", len(tarArchive))
-	fmt.Println("files to delete:", filesToDelete)
-	fmt.Println("cmds:", commands)
+	err := s.rmFiles(ctx, containerId, filesToDelete)
+	if err != nil {
+		return err
```
Comment (maiamcc) [inline internal/synclet/synclet.go]: maybe worth attaching metadata to these errs (i.e. the func that threw them)? but 🤷‍♀️

File: internal/synclet/synclet.go (hunk 1 above)
Comment (maiamcc) [inline internal/synclet/synclet.go]: add a todo to dedupe this function

File: internal/synclet/synclet.go (hunk 2 above)
Comment (landism) [reply internal/synclet/synclet.go]: Mostly to minimize edits/diffs when adding more calls or reordering calls (in that one wouldn't have to change `return s.dcli.CopyTo` to `err = s.dcli`...`if err != nil`), for the same reason some languages allow trailing commas after the last item in a list, but I don't know what's idiomatic go and don't really feel strongly either way.

File: internal/synclet/synclet.go (hunk 6 above)
Comment (landism) [reply internal/synclet/synclet.go]: whoops, thanks

File: internal/synclet/synclet.go (hunk 2 above)
Comment (maiamcc) [reply internal/synclet/synclet.go]: eh. fair. doesn't really matter i guess.

**Pull request merged**